Sources/StreamChat/WebSocketClient/EventMiddlewares/ChannelReadUpdaterMiddleware.swift

```diff
@@ -50,36 +56,104 @@ struct ChannelReadUpdaterMiddleware: EventMiddleware {
+            return log.error("Channel read is missing", subsystems: .webSocket)
         }
-
-        // Own messages don't increase unread count
-        guard let currentUserId = session.currentUser?.user.id, currentUserId != message.user.id else {
-            return
+        
+        if let skipReason = unreadCountUpdateSkippingReason(
+            currentUser: currentUser,
+            channelRead: channelRead,
+            message: message
+        ) {
+            return log.debug(
+                "Message \(message.id) does not increment unread counts for channel \(cid): \(skipReason)",
+                subsystems: .webSocket
+            )
         }
-
-        // Try to get the existing channel read for the current user
-        if let read = session.loadChannelRead(cid: cid, userId: currentUserId) {
-            if message.createdAt > read.lastReadAt ?? Date.distantPast {
-                read.unreadMessageCount += 1
-            }
-        } else {
-            log.info(
-                "Can't increase unread count for \(cid) because the `ChannelReadDTO` for " +
-                    "the current user doesn't exist."
+        
+        log.debug(
+            "Message \(message.id) increments unread counts for channel \(cid)",
+            subsystems: .webSocket
+        )
+        
+        channelRead.unreadMessageCount += 1
+    }
+    
+    private func decrementUnreadCountIfNeeded(
+        event: MessageDeletedEventDTO,
+        session: DatabaseSession
+    ) {
+        guard let currentUser = session.currentUser else {
+            return log.error("Current user is missing", subsystems: .webSocket)
+        }
+        
+        guard let channelRead = session.loadChannelRead(cid: event.cid, userId: currentUser.user.id) else {
+            return log.error("Channel read is missing", subsystems: .webSocket)
+        }
+        
+        if let skipReason = !event.hardDelete
+            ? "Soft-deleted messages do not affect unread counts"
+            : unreadCountUpdateSkippingReason(
+                currentUser: currentUser,
+                channelRead: channelRead,
+                message: event.message
+            ) {
+            return log.debug(
+                "Message \(event.message.id) does not decrement unread coutns for \(event.message.user.id): \(skipReason)",
+                subsystems: .webSocket
             )
         }
+        
+        log.debug(
+            "Message \(event.message.id) decrements unread counts for channel \(event.cid)",
+            subsystems: .webSocket
+        )
+        
+        channelRead.unreadMessageCount = max(0, channelRead.unreadMessageCount - 1)
+    }
+        
+    private func unreadCountUpdateSkippingReason(
+        currentUser: CurrentUserDTO,
+        channelRead: ChannelReadDTO,
+        message: MessagePayload
+    ) -> String? {
+        guard channelRead.channel.mute == nil else {
+            return "Channel is muted"
```

Comment:
Wouldn't it be more scalable if we introduced a UnreadSkippingReason? Also, is this logic meant to be overridable? Could the customers provide their only logic, or we don't want that?

Reply:
This logic just follows what's on the backend and keeps local unread counts in sync with the remote so, yeah, that's not supposed to be customizable by an integrator.

Reply:
Actually, an integrator will have control over some parts of this logic by specifying flags on app-level config (e.g. if they want to increment unread count when the channel is muted) and we will respect those flags